src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs

```diff
@@ -364,21 +364,22 @@ internal static unsafe int IndexOfNullByte(ref byte searchSpace)
             {
                 lengthToExamine -= 8;
 
-                if (uValue == Unsafe.AddByteOffset(ref searchSpace, offset))
+                ref byte current = ref Unsafe.AddByteOffset(ref searchSpace, offset);
+                if (uValue == current)
                     goto Found;
-                if (uValue == Unsafe.AddByteOffset(ref searchSpace, offset + 1))
+                if (uValue == Unsafe.AddByteOffset(ref current, 1))
                     goto Found1;
-                if (uValue == Unsafe.AddByteOffset(ref searchSpace, offset + 2))
+                if (uValue == Unsafe.AddByteOffset(ref current, 2))
                     goto Found2;
-                if (uValue == Unsafe.AddByteOffset(ref searchSpace, offset + 3))
+                if (uValue == Unsafe.AddByteOffset(ref current, 3))
                     goto Found3;
-                if (uValue == Unsafe.AddByteOffset(ref searchSpace, offset + 4))
+                if (uValue == Unsafe.AddByteOffset(ref current, 4))
                     goto Found4;
-                if (uValue == Unsafe.AddByteOffset(ref searchSpace, offset + 5))
+                if (uValue == Unsafe.AddByteOffset(ref current, 5))
                     goto Found5;
-                if (uValue == Unsafe.AddByteOffset(ref searchSpace, offset + 6))
+                if (uValue == Unsafe.AddByteOffset(ref current, 6))
                     goto Found6;
-                if (uValue == Unsafe.AddByteOffset(ref searchSpace, offset + 7))
+                if (uValue == Unsafe.AddByteOffset(ref current, 7))
```

Comment:
Do these changes actually help?  On coreclr at least it's not clear to me just from the diff that it's a net win.
https://www.diffchecker.com/du4Cevvb
Can you share throughput numbers?

Reply:
There shouldn't be any improvement on coreclr. I wouldn't be surprised if the generated code is more or less identical, since the JIT might make use of common subexpression elimination optimization. Also the code paths tweaked in this PR are for non vectorized cases, which are mostly not hit at all with coreclr.

The BCL change is intended for less optimized compilers to be able to generate good quality code.

Reply:
> There shouldn't be any improvement on coreclr.

I'm more concerned it might regress.  At least according to the diff I linked above, the new asm is a bit longer and appears to use a different addressing mode.  I don't have a good sense for what that might mean in terms of throughput.

> Also the code paths tweaked in this PR are for non vectorized cases, which are mostly not hit at all with coreclr.

Aren't they used for shorter inputs that don't have enough data to fill a vector?

Reply:
The local perf numbers on some benchmarks are unclear, I'm waiting for a full performance run.